### PR TITLE
Gh 504/client gateway integration

### PR DESCRIPTION
**Pull request by MouazAlzahabi**

### CODE REVIEW

File: Multisig/UI/Transactions/TransactionCell/CustomTransactionCellView.swift [hunk 1]
```diff
@@ -13,16 +13,16 @@ struct CustomTransactionCellView: View {
...
         HStack {
             Image("ico-custom-tx")
-            AddressCell(address: transaction.address, style: .shortAddressNoShare)
+            AddressCell(address: transaction.to, style: .shortAddressNoShare)
 
             Spacer()
 
             VStack {
-                Text("\(transaction.amount) \(transaction.tokenSymbol)")
+                Text(transaction.amount + " \(transaction.tokenSymbol)")
```
Comment (sche) [inline Multisig/UI/Transactions/TransactionCell/CustomTransactionCellView.swift]: why like this?

File: Multisig/UI/Transactions/TransactionCell/SettingsChangeTransactionCell/SettingsChangeTransactionCellView.swift [hunk 2]
```diff
@@ -15,17 +15,17 @@ struct SettingsChangeTransactionCellView: View {
     }
 
     var contentView: some View {
-        let setFallbackHandler = transaction.method as? MethodRegistry.GnosisSafeSettings.SetFallbackHandler
-        let enableModule = transaction.method as? MethodRegistry.GnosisSafeSettings.EnableModule
-        let disableModule = transaction.method as? MethodRegistry.GnosisSafeSettings.DisableModule
+        let setFallbackHandler = transaction.info as? SetFallbackHandlerSettingsChangeTransactionSammaryInfo
```
Comment (sche) [inline Multisig/UI/Transactions/TransactionCell/SettingsChangeTransactionCell/SettingsChangeTransactionCellView.swift]: Spelling: 'S_u_mmary'

File: Multisig/UI/Transactions/TransactionDetails/TransactionDetailsViewModel.swift [hunk 3]
```diff
@@ -11,57 +11,74 @@ import SwiftUI
...
-        Just(hash)
-            .compactMap { $0 }
-            .receive(on: DispatchQueue.global())
-            .tryMap { hash -> TransactionViewModel in
-                let transaction = try App.shared.safeTransactionService.transaction(hash: hash)
-                guard let safe = transaction.safe?.address else {
-                    throw Failure.safeInfoMissing
+        if let hash = hash {
+            Just(hash)
+                .compactMap { $0 }
```
Comment (sche) [inline Multisig/UI/Transactions/TransactionDetails/TransactionDetailsViewModel.swift]: is this compactMap needed?

File: Multisig/UI/Transactions/TransactionDetails/TransactionDetailsViewModel.swift [hunk 4]
```diff
@@ -11,57 +11,74 @@ import SwiftUI
...
+                    self.isRefreshing = false
+                }, receiveValue:{ [weak self] transaction in
+                    guard let `self` = self else { return }
+                    self.transactionDetails = transaction
+                    self.errorMessage = nil
+                })
+                .store(in: &subscribers)
+        } else if let id = id {
+            Just(id)
+            .receive(on: DispatchQueue.global())
```
Comment (sche) [inline Multisig/UI/Transactions/TransactionDetails/TransactionDetailsViewModel.swift]: tabulation

**Review by sche — APPROVED** [review]

File: Multisig/Data/Services/Safe Client Gateway Service/TransactionSummaryListRequest.swift [hunk 5]
```diff
@@ -14,7 +14,7 @@ struct TransactionSummaryListRequest: JSONRequest {
     var urlPath: String {
         "/v1/safes/\(safeAddress)/transactions"
     }
-    typealias ResponseType = Page<TransactionSummary>
+    typealias ResponseType =  Page<TransactionSummary>
```
Comment (DmitryBespalov) [inline Multisig/Data/Services/Safe Client Gateway Service/TransactionSummaryListRequest.swift]: extra space after `=`

File: Multisig/UI/Transactions/TransactionDetails/TransactionDetailsViewModel.swift [hunk 6]
```diff
@@ -11,57 +11,74 @@ import SwiftUI
...
+                    self.isLoading = false
+                    self.isRefreshing = false
+                }, receiveValue:{ [weak self] transaction in
+                    guard let `self` = self else { return }
+                    self.transactionDetails = transaction
+                    self.errorMessage = nil
+                })
+                .store(in: &subscribers)
+        } else if let id = id {
+            Just(id)
```
Comment (DmitryBespalov) [inline Multisig/UI/Transactions/TransactionDetails/TransactionDetailsViewModel.swift]: These two branches are very look-alike, please extract the copy-pasted part into a separate method that would accept a closure parameter that you can execute inside the tryMap

File: Multisig/UI/Transactions/TransactionListViewModel/CreationTransactionViewModel.swift [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+
+        factoryUsed = creattionTransactionInfo.factory?.address.checksummed
+    }
+
+    class func isValid(info: TransactionInfo) -> Bool {
+        info is CreationTransactionInfo
+    }
+
+    override class func viewModels(from tx: TransactionSummary) -> [TransactionViewModel] {
+        guard isValid(info: tx.txInfo) else {
```
Comment (DmitryBespalov) [inline Multisig/UI/Transactions/TransactionListViewModel/CreationTransactionViewModel.swift]: could be replaced with one-liner

```
isValid(info: tx.txInfo) ? [CreationTransactionViewModel(tx)] : []
```

The same one-liner could be used in teh CustomTransactionSummaryViewModel

File: Multisig/UI/Transactions/TransactionListViewModel/SettingChangeTransactionViewModel.swift [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        title = settingsChangeTransactionInfo.dataDecoded.method
+        self.info = settingsChangeTransactionInfo.settingsInfo
+    }
+
+    class func isValid(info: TransactionInfo) -> Bool {
+        info is SettingsChangeTransactionInfo
+    }
+    
+    override class func viewModels(from tx: TransactionSummary) -> [TransactionViewModel] {
+        guard isValid(info: tx.txInfo) else {
```
Comment (DmitryBespalov) [inline Multisig/UI/Transactions/TransactionListViewModel/SettingChangeTransactionViewModel.swift]: here, too, can be just one ternary operator

File: Multisig/UI/Transactions/TransactionListViewModel/TransactionViewModel+Factory.swift [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        [
+            TransferTransactionViewModel.self,
+            SettingChangeTransactionViewModel.self,
+            ChangeImplementationTransactionViewModel.self,
+            CustomTransactionViewModel.self,
+            CreationTransactionViewModel.self
+        ]
+        .map { $0.viewModels(from: tx) }
+        .first { !$0.isEmpty }
+        ?? []
```
Comment (DmitryBespalov) [inline Multisig/UI/Transactions/TransactionListViewModel/TransactionViewModel+Factory.swift]: the only thing that is different is the input parameter and the viewModels() variant, please remove the copy-pasted duplication by extracting the common part into a separate configurable method

File: Multisig/UI/Transactions/TransactionListViewModel/TransferTransactionViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    class func isValid(info: TransactionInfo) -> Bool {
+        info is TransferTransactionInfo
+    }
+
+    override class func viewModels(from tx: TransactionSummary) -> [TransactionViewModel] {
+        guard isValid(info: tx.txInfo) else {
+            return []
+        }
+
+        return [TransferTransactionViewModel(tx)]
```
Comment (DmitryBespalov) [inline Multisig/UI/Transactions/TransactionListViewModel/TransferTransactionViewModel.swift]: here, too, can be replaced by ternary operator

**Review by DmitryBespalov — CHANGES_REQUESTED** [review]
Big effort! 👍 

Some code duplication in several places and minor changes need to be corrected

**Review by DmitryBespalov — APPROVED** [review]

**Review by DmitryBespalov — APPROVED** [review]

**Pull request merged**